### PR TITLE
Add Geonames service and location validation steps

### PR DESCRIPTION
**Pull request by mcamarad**
# Why
Currently a lot of the location we retrieve from the geocoder service are wrong. Therefore we needed a validation step that compared and ensured that a location can be found in at least 2 location services.

# What
- Integration of geonames service into geocoder module and additional validation steps
- New unit tests
- Updated docs

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -8,16 +8,22 @@ This repository contains the code for the Geocoder module, that provides coordin
...
+- `geonames_api_endpoint`: api endpoint for the `Geonames` location service.
 - `url_api_endpoint`: api endpoint for the `Photon` geocoder.
 - `url_reverse_endpoint`: api endpoint for the `Photon` decoder (to get location from coordinates, not implemented for future use).
 - `lang`: parameter to specify the language of the returned results, default is `"en"`
 - `osm_keys`: list of filters on the geocoder results, based on Open Stree Map features https://wiki.openstreetmap.org/wiki/Map_features. Default is `place`
 
+### Installing module
+
+Remember to install the module, before running tests or you won't be able to test changes. Use `pip3 install -e .` to ensure that happens.
+
```
Comment (ericanand) [inline README.md:20]: Just a suggestion - maybe we could call this section `Running the unit tests` for something like that, and give people a one-liner on how to run the tests (and what env variables to use)?

```suggestion
### Running the unit tests

Remember to install the module, before running tests or you won't be able to test your changes as the `pytest` command will run against the package. The `pip3 install -e .` will ensure this happens.

```shell
export PHOTON_SERVER=test 	# Set dummy environment variable when running tests
export GEONAMES_SERVER=test # Set dummy environment variable when running tests

pip3 install -e . && pytest -vv	# Run the tests while updating the package with latest changes
```

```

File: codeship-steps.yml [hunk 2]
```diff
@@ -0,0 +1,6 @@
+- name: Run unit tests
+  type: parallel
+  steps:
+    - name: tests
+      service: run-unit-tests
+      command: pytest -vv
```
Comment (ericanand) [inline codeship-steps.yml:6]: 😍

File: geocoder_module/geocoder.py [hunk 3]
```diff
@@ -36,6 +36,7 @@ def __init__(self) -> None:
         """
 
         self.config = {
+            "geonames_api_endpoint": "/locations/?",
```
Comment (ericanand) [inline geocoder_module/geocoder.py:39]: Small nit - any reason why this endpoint has a backslash at the end, but the others don't? More a question out of interest 😁

File: geocoder_module/geocoder.py [hunk 4]
```diff
@@ -87,7 +88,60 @@ def __init__(self) -> None:
...
+            url_api = (
+                os.environ["GEONAMES_SERVER"] + self.config["geonames_api_endpoint"]
+            )
+
+            response = requests.get(
+                url_api,
+                params={"country": country.lower(), "local_location": location.lower()},
+            )
+            response = response.json()
+            print(f"Geonames response : {response}")
```
Comment (ericanand) [inline geocoder_module/geocoder.py:125]: We can remove this print statement now 😄 
```suggestion
```

File: geocoder_module/geocoder.py [hunk 5]
```diff
@@ -122,6 +175,7 @@ def get_location_info(
                 },
             )
             response = response.json()
+            print(f"Geocoder response : {response}")
```
Comment (ericanand) [inline geocoder_module/geocoder.py:178]: Can remove this too 😄 
```suggestion
```

File: requirements.txt [hunk 6]
```diff
@@ -1 +1,2 @@
 requests==2.25.1
+pytest-asyncio==0.15.1
```
Comment (ericanand) [inline requirements.txt:2]: Do we end up using this? Can likely remove this dependency I believe.
```suggestion
```

File: tests/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,20 @@
+FROM python:3.9
+
+WORKDIR /usr/src/app
+
+# set python path to include working directory
+ENV PYTHONPATH="${PYTHONPATH}:/usr/src/app"
+
+ENV PHOTON_SERVER=test
+ENV GEONAMES_SERVER=test
```
Comment (ericanand) [inline tests/Dockerfile:9]: Instead of adding these here, shall we add it at runtime in the Codeship services definition? For example: https://github.com/lifebit-ai/nlp-sample-creation/blob/main/codeship-services.yml#L5-L8

That way we're are injecting the variables at runtime, and not baking them into our docker image. It also keeps all of our unit tests across our services consistent in format.

File: tests/test_get_location.py [hunk 8]
```diff
@@ -0,0 +1,205 @@
+from unittest.mock import Mock, patch
+import pytest
+
+from geocoder_module.geocoder import Geocoder
+
+geocoder = Geocoder()
+
+
+@patch("requests.get")
+def test_get_geocoder_info_returns_a_location_when_queried(
```
Comment (ericanand) [inline tests/test_get_location.py:10]: Fantastic work on these unit tests! Big fan 🎉

**Review by ericanand — COMMENTED** [review]
Great stuff - some small things to fix but on the whole really good 👍

File: geocoder_module/geocoder.py (hunk 3 above)
Comment (mcamarad) [reply geocoder_module/geocoder.py:39]: Not really that's just how the endpoint looked like it in the browser. You can either use the / or not 😄

File: tests/test_get_location.py (hunk 8 above)
Comment (mcamarad) [reply tests/test_get_location.py:10]: Thanks a lot for the great comments! 😄

**Comment by mcamarad** [comment]
Ready for round 2️⃣  🛎️

**Review by ericanand — APPROVED** [review]
Looks good to me 👍

**Comment by mcamarad** [comment]
I had to add more validation logic to cover some edge cases in the geocode.py script as a result of some test in the samples creation. Could you please give it another look @ericanand ?

**Comment by ericanand** [comment]
I think the changes look fine - tbh as I haven't run this service locally I couldn't tell you for sure, but if you had some time would a couple of unit test to cover the new logic (the `double_check_countries` function) help give you some confidence here?

It's turned into quite a big function, personally I'd break it down to make it more testable - however this is only if you have the time.

**Comment by ericanand** [comment]
>I need to add the GEONAMES_SERVER to all repos using the geocoder service as an env variable, this means sample-creation and outbreak detection.

Ah this is true - I also need to raise an infrastructure PR to add this to our environment variables, else the production services will fail.

If you update the sample creation service, I'll update the outbreak detection one (I'm going to raise a PR for this service anyway).

**Comment by ericanand** [comment]
[This PR](https://github.com/lifebit-ai/lifebit-infrastructure-live/pull/626/files), will need to be merged in before we merge this one 🙏

**Comment by ericanand** [comment]
Great stuff with the unit tests - looks like you thoroughly tried different use cases which is great to see 👌

**Review by ericanand — APPROVED** [review]
